### PR TITLE
session, planner: Add a factor getter for networkFactor/scanFactor/descScanFactor/seekFactor

### PR DESCRIPTION
**Pull request by mmyj**
### What problem does this PR solve?

Problem Summary:

Add the factor getter for networkFactor/scanFactor/descScanFactor/seekFactor. The getter will return 0, when the data source table is a temporary table.

### What is changed and how it works?

How it Works:

The getter will return 0, when the `model.TableInfo` is a temporary table.

### Related changes

- N/A

### Check List 

Tests 

- No code

Side effects

- N/A

### Release note 
- Add a factor getter for networkFactor/scanFactor/descScanFactor/seekFactor

### CODE REVIEW

File: planner/core/task.go [hunk 1]
```diff
@@ -151,18 +151,19 @@ func (t *copTask) finishIndexPlan() {
 	t.indexPlanFinished = true
 	sessVars := t.indexPlan.SCtx().GetSessionVars()
 	// Network cost of transferring rows of index scan to TiDB.
-	t.cst += cnt * sessVars.NetworkFactor * t.tblColHists.GetAvgRowSize(t.indexPlan.SCtx(), t.indexPlan.Schema().Columns, true, false)
+	t.cst += cnt * sessVars.GetNetworkFactor(nil) * t.tblColHists.GetAvgRowSize(t.indexPlan.SCtx(), t.indexPlan.Schema().Columns, true, false)
```
Comment (djshow832) [inline planner/core/task.go:154]: For temporary tables, the network factor is 0. Is it possible to pass a `TableInfo`?

File: planner/implementation/datasource.go [hunk 2]
```diff
@@ -74,7 +74,7 @@ func (impl *TableReaderImpl) CalcCost(outCount float64, children ...memo.Impleme
 	reader := impl.plan.(*plannercore.PhysicalTableReader)
 	width := impl.tblColHists.GetAvgRowSize(impl.plan.SCtx(), reader.Schema().Columns, false, false)
 	sessVars := reader.SCtx().GetSessionVars()
-	networkCost := outCount * sessVars.NetworkFactor * width
+	networkCost := outCount * sessVars.GetNetworkFactor(nil) * width
```
Comment (djshow832) [inline planner/implementation/datasource.go:77]: Why nil?

File: planner/core/task.go (hunk 1 above)
Comment (mmyj) [reply planner/core/task.go]: L161 will recount cost using TableInfo.

File: planner/implementation/datasource.go (hunk 2 above)
Comment (mmyj) [reply planner/implementation/datasource.go:79]: TableReaderImpl don't have tableInfo property, so using nil to replace it. I will create another PR to fix it.

File: planner/core/task.go [hunk 3]
```diff
@@ -151,18 +151,21 @@ func (t *copTask) finishIndexPlan() {
...
 	// Network cost of transferring rows of index scan to TiDB.
-	t.cst += cnt * sessVars.NetworkFactor * t.tblColHists.GetAvgRowSize(t.indexPlan.SCtx(), t.indexPlan.Schema().Columns, true, false)
+	cst := t.cst + cnt*sessVars.GetNetworkFactor(nil)*t.tblColHists.GetAvgRowSize(t.indexPlan.SCtx(), t.indexPlan.Schema().Columns, true, false)
 
 	if t.tablePlan == nil {
+		t.cst = cst
 		return
 	}
+	ts := t.tablePlan.(*PhysicalTableScan)
+	t.cst += cnt * sessVars.GetNetworkFactor(ts.Table) * t.tblColHists.GetAvgRowSize(t.indexPlan.SCtx(), t.indexPlan.Schema().Columns, true, false)
```
Comment (djshow832) [inline planner/core/task.go:161]:
```suggestion
	var tblInfo model.TableInfo
	var ts *PhysicalTableScan
	if t.tablePlan != nil {
		ts = t.tablePlan.(*PhysicalTableScan)
		tblInfo = ts.Table
	}
	t.cst += cnt * sessVars.GetNetworkFactor(tblInfo) * t.tblColHists.GetAvgRowSize(t.indexPlan.SCtx(), t.indexPlan.Schema().Columns, true, false)

	if t.tablePlan == nil {
		return
	}
```

File: planner/core/task.go [hunk 4]
```diff
@@ -150,19 +151,27 @@ func (t *copTask) finishIndexPlan() {
 	cnt := t.count()
 	t.indexPlanFinished = true
 	sessVars := t.indexPlan.SCtx().GetSessionVars()
+	var (
+		tableInfo *model.TableInfo
+		ts        *PhysicalTableScan
+	)
+	if t.tablePlan != nil {
+		ts = t.tablePlan.(*PhysicalTableScan)
```
Comment (djshow832) [inline planner/core/task.go:159]:
```suggestion
	var tableInfo *model.TableInfo
	if t.tablePlan != nil {
		ts := t.tablePlan.(*PhysicalTableScan)
```

**Comment by djshow832** [comment]
@qw4990  Do you think it's OK to set factors to 0? Is it necessary to set scan factor to 0?

**Review by djshow832 — APPROVED** [review]

**Review by qw4990 — APPROVED** [review]

**Pull request merged**